File: java/persistent_lazy_rbtree/src/PersistentLazyRBTree.java
package persistent_lazy_rbtree.src;

enum Color {
    RED,
    BLACK,
}

interface Node<T> {

    class Pair<T> {
        final Node<T> e0, e1;

        Pair(Node<T> e0, Node<T> e1) {
            this.e0 = e0;
            this.e1 = e1;
        }
    }

    Color color();

    int rank();

    int size();

    Node<T> left();

    Node<T> right();

    T index(int index);

    Pair<T> split(int index);

    static <T> Node<T> merge(Node<T> left, Node<T> right) {
        if (left.rank() < right.rank()) {
            left = merge(left, right.left());
            if (left.color() == Color.RED && left.left().color() == Color.RED && right.color() == Color.BLACK) {
                if (right.right().color() == Color.BLACK) {
                    return new Tree<T>(Color.BLACK, left.left(), new Tree<T>(Color.RED, left.right(), right.right()));
                } else {
                    return new Tree<T>(Color.RED, new Tree<T>(Color.BLACK, left.left(), left.right()),
                            new Tree<T>(Color.BLACK, right.right().left(), right.right().right()));
                }
            } else {
                return new Tree<T>(right.color(), left, right.right());
            }
        } else if (left.rank() > right.rank()) {
            right = merge(left.right(), right);
            if (left.color() == Color.BLACK && right.right().color() == Color.RED && right.color() == Color.RED) {
                if (left.left().color() == Color.BLACK) {
                    return new Tree<T>(Color.BLACK, new Tree<T>(Color.RED, left.left(), right.left()), right.right());
                } else {
                    return new Tree<T>(Color.RED, new Tree<T>(Color.BLACK, left.left().left(), left.left().right()),
                            new Tree<T>(Color.BLACK, right.left(), right.right()));
                }
            } else {
                return new Tree<T>(left.color(), left.left(), right);
            }
        } else {
            return new Tree<T>(Color.RED, left, right);
        }
    }
}

class InvalidMethodException extends RuntimeException {
    InvalidMethodException() {
        super();
    }
}

class Leaf<T> implements Node<T> {
    final T val;

    Leaf(T val) {
        this.val = val;
    }

    @Override
    public Color color() {
        return Color.BLACK;
    }

    @Override
    public int rank() {
        return 0;
    }

    @Override
    public int size() {
        return 1;
    }

    @Override
    public Node<T> left() {
        throw new InvalidMethodException();
    }

    @Override
    public Node<T> right() {
        throw new InvalidMethodException();
    }

    @Override
    public T index(int index) {
        return val;
    }

    @Override
    public Pair<T> split(int index) {
        throw new InvalidMethodException();
    }
}

class Tree<T> implements Node<T> {
    final Color color;
    final int rank, size;
    final Node<T> left, right;

    Tree(Color color, Node<T> left, Node<T> right) {
        this.color = color;
        this.rank = left.rank() + (left.color() == Color.BLACK ? 1 : 0);
        this.size = left.size() + right.size();
        this.left = left;
        this.right = right;
    }

    @Override
    public Color color() {
        return Color.BLACK;
    }

    @Override
    public int rank() {
        return 0;
    }

    @Override
    public int size() {
        return 1;
    }

    @Override
    public Node<T> left() {
        throw new InvalidMethodException();
    }

    @Override
    public Node<T> right() {
        throw new InvalidMethodException();
    }

    @Override
    public T index(int index) {
        if (index < left.size()) {
            return left.index(index);
        } else {
            return right.index(index - left.size());
        }
    }

    @Override
    public Pair<T> split(int index) {
        if (index < left.size()) {
            Pair<T> _left = left.split(index);
            return new Pair<T>(_left.e0, Node.merge(_left.e1, right));
        } else if (index > left.size()) {
            Pair<T> _right = right.split(index - left.size());
            return new Pair<T>(Node.merge(left, _right.e0), _right.e1);
        } else {
            return new Pair<T>(left, right);
        }
    }
}

public class PersistentLazyRBTree<T> {
    public class Pair {
        final PersistentLazyRBTree<T> e0, e1;

        Pair(PersistentLazyRBTree<T> e0, PersistentLazyRBTree<T> e1) {
            this.e0 = e0;
            this.e1 = e1;
        }
    }

    private final Node<T> root;

    private PersistentLazyRBTree(Node<T> root) {
        this.root = root;
    }

    public PersistentLazyRBTree() {
        this.root = null;
    }

    public int size() {
        if (root == null) {
            return 0;
        } else {
            return root.size();
        }
    }

    public Pair split(int index) {
        if (index == 0) {
            return new Pair(new PersistentLazyRBTree<T>(), this);
        } else if (index == size()) {
            return new Pair(this, new PersistentLazyRBTree<T>());
        } else {
            Node.Pair<T> result = root.split(index);
            return new Pair(new PersistentLazyRBTree<T>(result.e0), new PersistentLazyRBTree<T>(result.e1));
        }
    }

    public static <T> PersistentLazyRBTree<T> merge(PersistentLazyRBTree<T> left, PersistentLazyRBTree<T> right) {
        if (left.root == null) {
            return right;
        } else if (right.root == null) {
            return left;
        } else {
            Node<T> root = Node.merge(left.root, right.root);
            return new PersistentLazyRBTree<T>(new Tree<T>(Color.BLACK, root.left(), root.right()));
        }
    }
}

File: py/persistent_lazy_rbtree/persistent_lazy_rbtree.py
from enum import Enum, auto


class Color(Enum):
    RED = auto()
    BLACK = auto()


class Leaf:
    def __init__(self, val):
        self.__val = val

    def __len__(self):
        return 1

    def __getitem__(self, _):
        return self.__val

    def color(self):
        return Color.BLACK

    def rank(self):
        return 0


class Tree:
    def __init__(self, color, left, right):
        self.__color = color
        self.__len = len(left) + len(right)
        self.__rank = left.rank() + (1 if color == Color.BLACK else 0)
        self.left = left
        self.right = right

    def __len__(self):
        return self.__len

    def __getitem__(self, index):
        if index < len(self.left):
            return self.left[index]
        else:
            return self.right[index - len(self.left)]

    def color(self):
        return self.__color

    def rank(self):
        return self.__rank

    @staticmethod
    def merge(left, right):
        if left.rank() < right.rank():
            left = Tree.merge(left, right.left)
            if left.color() == Color.RED and left.left.color() == Color.RED and right.color() == Color.BLACK:
                if right.right.color() == Color.BLACK:
                    return Tree(Color.BLACK, left.left, Tree(Color.RED, left.right, right.right))
                else:
                    return Tree(Color.RED, Tree(Color.BLACK, left.left, left.right), Tree(Color.RED, right.right.left, right.right.right))
            else:
                return Tree(right.color(), left, right.right)
        elif left.rank() > right.rank():
            right = Tree.merge(left.right, right)
            if left.color() == Color.BLACK and right.right.color() == Color.RED and right.color() == Color.RED:
                if left.left.color():
                    return Tree(Color.BLACK, Tree(Color.RED, left.left, right.left), right.right)
                else:
                    return Tree(Color.RED, Tree(Color.Black, left.left.left, left.left.right), Tree(Color.BLACK, right.left, right.right))
            else:
                return Tree(left.color(), left.left, right)
        else:
            return Tree(Color.RED, left, right)

    def split(self, index):
        if index < len(self.left):
            left_left, left_right = self.left.split(index)
            return left_left, Tree.merge(left_right, self.right)
        elif index > len(self.left):
            right_left, right_right = self.right.split(index - len(self.left))
            return Tree.merge(self.left, right_left), right_right
        else:
            self.left, self.right


class PersistentLazyRBTree:
    def __set(self, root):
        self.__root = root

    def __getitem__(self, index):
        return self.__root[index]

    @staticmethod
    def merge(left, right):
        if left.__root == None:
            return right
        elif right.__root == None:
            return left
        else:
            root = Tree.merge(left.__root, right.__root)
            return PersistentLazyRBTree().__set(Tree(Color.BLACK, root.left, root.right))

    def split(self, index):
        if index == 0:
            return (PersistentLazyRBTree(), self)
        elif index == len(self):
            return (self, PersistentLazyRBTree())
        else:
            left, right = self.__root.split(index)
            return PersistentLazyRBTree().__set(left), PersistentLazyRBTree().__set(right)

    def insert(self, index, val):
        left, right = self.split(index)
        return PersistentLazyRBTree.merge(PersistentLazyRBTree.merge(left, PersistentLazyRBTree().__set(Leaf(val))), right)

    def erase(self, index):
        left, right = self.split(index)
        _, right = right.split(1)
        return PersistentLazyRBTree.merge(left, right)
